### ISSUE TITLE
Major compaction shouldn't block compaction on newly created files

### ISSUE THREAD

**Issue by raphaelsc** [scylladb/scylla#10961]
Major compaction is serialized with regular compaction, so if it takes a long time, L0 sstables can accumulate indefinitely, leading to OOM / terrible read latency due to unbounded read amplification.

let's fix it by limiting the serialization to the selection of candidates phase, so once major starts running, it's only serialized to another major and regular will be able to run on files created after that.

**Comment by raphaelsc** [comment]
FYI: @bhalevy

**Comment by raphaelsc** [comment]
> > Major compaction is serialized with regular compaction, so if it takes a long time, L0 sstables can accumulate indefinitely, leading to OOM / terrible read latency due to unbounded read amplification.
> 
> I was under the impression (do we have the current algorithm documented anywhere?) that we prevent concurrent compactions depending on the size - so a really big compaction like major compaction would only prevent other really big compactions - but not tiny compactions. I guess this wasn't the full story and you also have a major-compaction serialization mutex, but I wonder if it's really needed:

We had a rwlock for serializing regular and major on behalf of a table T, but that's overkill and we need to allow regular to proceed in parallel to major, which will be either accepted or rejected based on the size serialization logic. Agree we need better documentation, will write one.

> 
> > let's fix it by limiting the serialization to the selection of candidates phase, so once major starts running, it's only serialized to another major and regular will be able to run on files created after that.
> 
> I don't think major compaction needs to be "serialized" in the sense that if you run "nodetool compact" and there is a major compaction already running, it will wait for that compaction to finish and then start another major compaction. Instead, it should tell you that there is already another major compaction running, and fail immediately. No?

If you run nodetool compaction on a keyspace, then we want major to happen on a single table at a time, to avoid huge space requirement. But for a single table, or if you run nodetool compact on an entire keyspace, major is serialized. Agree we could possibly ignore / fail a second request if there's an ongoing major, but we'd also be changing an API here. What if an user really wants a second major to take place while the current one is almost done? with incremental compaction, major replaces input files with output ones incrementally.

> 
> By the way, what happens if a non-major compaction is currently running? Do you wait for _that_ before starting the major compaction? Even if it may take another hour? Or do you stop it in the middle and start the major compaction instead?

we have two options. kill the ongoing compaction or wait for it. today we wait for it for selecting the candidates. but that's something we have to think about carefully. we cannot kill user-triggered compaction like cleanup to run major, as the application which ran cleanup automatically may break. Again, that's an API change which we have to be careful if changing.

**Comment by bhalevy** [comment]
> we have two options. kill the ongoing compaction or wait for it. today we wait for it for selecting the candidates. but that's something we have to think about carefully.

The regular compaction we wait for is essentially doing part of the work major compaction will need to do anyway.
It might make sense to stop it if it had only started working, but if it's already made enough progress it should be more efficient to let it finish.

I think that ultimately the best solution would be to have a per-table mode for major compaction and just let regular compaction consider it in its bucket selection algorithm so it would select the buckets in major compaction order, but also consider buckets that cross some threshold so to tend to the backlog.  It should keep going until all the sstables it started with are compacted and then set a value on a major_compaction_done promise that the api call can wait on.

**Comment by raphaelsc** [comment]
> > we have two options. kill the ongoing compaction or wait for it. today we wait for it for selecting the candidates. but that's something we have to think about carefully.
> 
> The regular compaction we wait for is essentially doing part of the work major compaction will need to do anyway. It might make sense to stop it if it had only started working, but if it's already made enough progress it should be more efficient to let it finish.
> 
> I think that ultimately the best solution would be to have a per-table mode for major compaction and just let regular compaction consider it in its bucket selection algorithm so it would select the buckets in major compaction order, but also consider buckets that cross some threshold so to tend to the backlog. It should keep going until all the sstables it started with are compacted and then set a value on a major_compaction_done promise that the api call can wait on.

major requires compacting all sstables together in the same compaction round. In my opinion, we shouldn't mix regular and major types. We can kill ongoing compaction if it's going to take lots of time.

**Comment by bhalevy** [comment]
> We can kill ongoing compaction if it's going to take lots of time.

Major doesn't know how long the ongoing compactions will take, and what would that buy?
Any work already done by regular compaction will have to be repeated by major anyhow, so why not just let it complete?

**Comment by avikivity** [comment]
We can insert a barrier that says: only allow decreasing-weight comparisons from now on, then launch the major compaction (on all sstables), and then allow compactions in parallel to the major compaction (since these would be on new files). This would allow invested work to continue, void tiny sstables building up, and converge.

But it's simpler to just abort running compactions. Major compaction is rare enough and it's better to strive for predictability and simplicity. It will also test our compaction abort.

**Comment by bhalevy** [comment]
> We can insert a barrier that says: only allow decreasing-weight comparisons from now on, then launch the major compaction (on all sstables), and then allow compactions in parallel to the major compaction (since these would be on new files). This would allow invested work to continue, void tiny sstables building up, and converge.
> 
> But it's simpler to just abort running compactions. 

There's is real need to, major gets a write lock which is a blocking barrier that waits on running regular compaction jobs and prevents new regular compaction jobs from starting.  Then it grabs all sstables and releases the lock to allow regular compaction for new sstables.

> Major compaction is rare enough and it's better to strive for predictability and simplicity. It will also test our compaction abort.

Aborting regular compaction is exercized on a regular basis when dropping tables in tests.

**Comment by avikivity** [comment]
Significant change, bug is fairly rare and involves rare operation (major compaction). Not backporting.
